### PR TITLE
Addition of position argument to TreeSequence.site()

### LINKED ISSUE #2234: Add position=None keyword only arg to TreeSequence.site() method
The ``TreeSequence.site()`` method currently takes a single argument, ``id``. It would be very handy for exploratory analysis to have a ``position=x`` argument.

We can use use np.binsearch on ``tables.sites.position`` to find it, and raise a KeyError if it doesn't exist.

@szhan Fancy picking this on up?

### PR DESCRIPTION
**Pull request by szhan**
Fixes #2234

### CODE REVIEW

**Comment by benjeffery** [comment]
@szhan Tag me here when you want a 👀

**Comment by jeromekelleher** [comment]
@szhan - it would be nice to get this one in for the next release. It mainly just needs some tests - do you want some pointers on how to do this?

**Comment by szhan** [comment]
I'm trying to find where the tests are for ``TreeSequence.site()``. I'm not seeing them under ``class TestTreeSequence`` in ``test_highlevel.py``. Are they elsewhere?

**Comment by jeromekelleher** [comment]
No, there's probably no specific testing for this method. I'd suggest adding a ``TestTreeSequenceGetSite`` method to tests/test_highlevel.py where you can add some simple tests.

**Comment by benjeffery** [comment]
> I'm trying to find where the tests are for `TreeSequence.site()`. I'm not seeing them under `class TestTreeSequence` in `test_highlevel.py`. Are they elsewhere?

`test_highlevel:2006`

**Comment by szhan** [comment]
I see a test for ``TreeSequence.sites()`` in which ``TreeSequence.site()`` get used, but not tests written specifically for ``TreeSequence.site()``.

File: python/tests/test_highlevel.py [hunk 1]
```diff
@@ -4261,3 +4261,24 @@ def test_macs_error(self):
             ValueError, match="macs output only supports single letter alleles"
         ):
             ts.to_macs()
+
+
+class TestTreeSequenceGetSite:
```
Comment (benjeffery) [inline python/tests/test_highlevel.py:4266]: The tests at test_highlevel.py:2006 are specifically for the site method. These tests are a good start, but I think adding 
```python
s3 = ts.site(position=site.postion)
assert s3 = site
```
would be an good thing to add too.

File: python/tests/test_highlevel.py [hunk 2]
```diff
@@ -4261,3 +4261,24 @@ def test_macs_error(self):
...
+    """
+    Tests for getting Site objects from a TreeSequence object
+    by specifying site id or position.
```
Comment (benjeffery) [inline python/tests/test_highlevel.py:4269]: These only do position though, so leave it at that?

File: python/tests/test_highlevel.py [hunk 3]
```diff
@@ -4261,3 +4261,24 @@ def test_macs_error(self):
...
+
+    def get_example(self):
+        tables = tskit.TableCollection(sequence_length=10)
+        tables.sites.add_row(position=3, ancestral_state="A")
+        tables.sites.add_row(position=5, ancestral_state="C")
+        tables.sites.add_row(position=7, ancestral_state="G")
+        tables.sites.add_row(position=8.3, ancestral_state="T")
+        return tables.tree_sequence()
+
+    @pytest.mark.parametrize("position", [3.0, 5.0, 7.0, 8.3])
```
Comment (benjeffery) [inline python/tests/test_highlevel.py:4280]: Would be good add tests for the `0` and `10` edge cases here.

File: python/tskit/trees.py [hunk 4]
```diff
@@ -5098,13 +5098,27 @@ def mutation(self, id_):
             metadata_decoder=self.table_metadata_schemas.mutation.decode_row,
         )
 
-    def site(self, id_):
+    def site(self, id_=None, position=None):
         """
         Returns the :ref:`site <sec_site_table_definition>` in this tree sequence
-        with the specified ID.
+        with either the specified ID or position.
```
Comment (benjeffery) [inline python/tskit/trees.py:5104]: An explanation of the scaling characteristics would be good here as the caller should be informed that this method when called with `position=` does a binary search of the position array. We don't want people to iterate through positions to get sites!

File: python/tskit/trees.py [hunk 5]
```diff
@@ -5098,13 +5098,27 @@ def mutation(self, id_):
...
 
         :rtype: :class:`Site`
         """
+        if id_ is None and position is None:
+            raise TypeError("Site id or position must be provided.")
```
Comment (benjeffery) [inline python/tskit/trees.py:5109]: Missing coverage in all the error states.

File: python/tskit/trees.py [hunk 6]
```diff
@@ -5098,13 +5098,27 @@ def mutation(self, id_):
...
+            if position < 0 or position >= self.sequence_length:
+                raise ValueError(
+                    "Position is beyond the coordinates defined by sequence length."
+                )
+            position = np.array(position)
+            if len(position.shape) > 0:
+                raise ValueError("Position must be provided as a scalar value.")
+            site_pos = self.tables.sites.position  # Sorted in ascending order
+            id_ = site_pos.searchsorted(position)
+            if site_pos[id_] != position:
```
Comment (benjeffery) [inline python/tskit/trees.py:5120]: Don't you end up getting a `KeyError` here if the position is greater than the last site? `searchsorted` will return one-past-the-end.

**Review by benjeffery — COMMENTED** [review]
Thanks @szhan!
Needs tests for the error states, some more detail on the docs and a fix for a possible KeyError.

File: python/tests/test_highlevel.py [hunk 7]
```diff
@@ -4261,3 +4261,23 @@ def test_macs_error(self):
...
+        tables.sites.add_row(position=3, ancestral_state="A")
+        tables.sites.add_row(position=5, ancestral_state="C")
+        tables.sites.add_row(position=7, ancestral_state="G")
+        return tables.tree_sequence()
+
+    @pytest.mark.parametrize("position", [3.0, 5.0, 7.0])
+    def test_site_position(self, position):
+        ts = self.example()
+        site = ts.site(position=position)
+        assert site.position == position
```
Comment (jeromekelleher) [inline python/tests/test_highlevel.py:4284]: This is a good start. So we also need:
```
def test_site_id_and_position(self):
    pass

def test_no_args(self):
    pass

@pytest.mark.parametrize(...)
def test_position_bad_type(self, bad_type):
    pass

@pytest.mark.parametrize("position", [-1, 10, 11]):
def test_position_out_of_bounds(self, position):
    pass

@pytest.mark.parametrize("position", [0, 2.999999, 3.000001, 4.9999, 5.00001, 9.99999]):
def test_bad_position(self, position):
    pass

File: python/tests/test_highlevel.py [hunk 8]
```diff
@@ -4261,3 +4261,23 @@ def test_macs_error(self):
...
+    """
+
+    def get_example():
+        tables = tskit.TableCollection(sequence_length=10)
+        tables.sites.add_row(position=3, ancestral_state="A")
+        tables.sites.add_row(position=5, ancestral_state="C")
+        tables.sites.add_row(position=7, ancestral_state="G")
+        return tables.tree_sequence()
+
+    @pytest.mark.parametrize("position", [3.0, 5.0, 7.0])
```
Comment (jeromekelleher) [inline python/tests/test_highlevel.py]: Also want to test different number types here, so something like
```
[3.0, 5.0, 7.0, 3, np.array([3], dtype=float)[0], np.array([3], dtype=int)[0], decimal.Decimal(3)]
```

**Review by jeromekelleher — COMMENTED** [review]
Good start, just needs more tests for all the tricky corner cases. There's probably more I haven't thought of here, like what happens when you pass a list of values, etc.

File: python/tskit/trees.py [hunk 9]
```diff
@@ -5098,13 +5098,27 @@ def mutation(self, id_):
             metadata_decoder=self.table_metadata_schemas.mutation.decode_row,
         )
 
-    def site(self, id_):
+    def site(self, id_=None, position=None):
```
Comment (jeromekelleher) [inline python/tskit/trees.py:5101]: Position should be keyword only, so need 
```
    def site(self, id_=None, *, position=None):
```

File: python/tskit/trees.py (hunk 6 above)
Comment (szhan) [reply python/tskit/trees.py:5120]: I have added this assertion in ``site`` before ``searchsorted``. It only works for discrete coordinates though.

```
if position < 0 or position >= self.sequence_length:
    raise ValueError("Position is beyond the coordinates defined by sequence length.")
```

Comment (benjeffery) [reply python/tskit/trees.py:5120]: I don't think that works even for discrete coords - say for example if you had zero sites!

File: python/tests/test_highlevel.py (hunk 7 above)
Comment (szhan) [reply python/tests/test_highlevel.py:4292]: It should only work for discrete positions, not continuous?

**Comment by szhan** [comment]
I keep running into this error when trying to ``git commit``:

```
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

python/tests/test_highlevel.py:44:1: F401 'curses.ACS_NEQUAL' imported but unused
python/tests/test_highlevel.py:52:1: F401 'python.tests.simplify.AncestorMap' imported but unused
```

So, I have been using ``SKIP=flake8`` for now.

File: python/tskit/trees.py (hunk 6 above)
Comment (szhan) [reply python/tskit/trees.py:5128]: I have added a test to check that a KeyError is returned in such an instance.

File: python/tests/test_highlevel.py (hunk 8 above)
Comment (szhan) [reply python/tests/test_highlevel.py]: I have added a check that position must be an integer or floating point. Is that too narrow?

File: python/tskit/trees.py (hunk 6 above)
Comment (benjeffery) [reply python/tskit/trees.py:5128]: I don't think `KeyError` is the right exception to pass back to the user here. Can you add a test for getting a site from a table with no sites also to exercise this part and check we get the right exception? (`ValueError`)

Comment (szhan) [reply python/tskit/trees.py:5128]: I think I have corrected it now.

File: python/tests/test_highlevel.py (hunk 8 above)
Comment (szhan) [reply python/tests/test_highlevel.py]: Ahhh, why there are so many types of numbers?

File: python/tests/test_highlevel.py (hunk 7 above)
Comment (szhan) [reply python/tests/test_highlevel.py:4303]: Could I do this to check whether ``position`` holds an integer?

```
a = 9.4
int(a) == a
a = 3.0
int(a) == a
```

File: python/tests/test_highlevel.py (hunk 8 above)
Comment (benjeffery) [reply python/tests/test_highlevel.py]: Checking types like this is usually an anti-pattern in python 

> I keep running into this error when trying to `git commit`:
> 
> ```
> flake8...................................................................Failed
> - hook id: flake8
> - exit code: 1
> 
> python/tests/test_highlevel.py:44:1: F401 'curses.ACS_NEQUAL' imported but unused
> python/tests/test_highlevel.py:52:1: F401 'python.tests.simplify.AncestorMap' imported but unused
> ```
> 
> So, I have been using `SKIP=flake8` for now.

@szhan Found that these lines had been added 
to his patch without his knowledge!

**Comment by benjeffery** [comment]
> I keep running into this error when trying to `git commit`:
> 
> ```
> flake8...................................................................Failed
> - hook id: flake8
> - exit code: 1
> 
> python/tests/test_highlevel.py:44:1: F401 'curses.ACS_NEQUAL' imported but unused
> python/tests/test_highlevel.py:52:1: F401 'python.tests.simplify.AncestorMap' imported but unused
> ```
> 
> So, I have been using `SKIP=flake8` for now.

Checking types like this is usually an anti-pattern in python 

> I keep running into this error when trying to `git commit`:
> 
> ```
> flake8...................................................................Failed
> - hook id: flake8
> - exit code: 1
> 
> python/tests/test_highlevel.py:44:1: F401 'curses.ACS_NEQUAL' imported but unused
> python/tests/test_highlevel.py:52:1: F401 'python.tests.simplify.AncestorMap' imported but unused
> ```
> 
> So, I have been using `SKIP=flake8` for now.

@szhan Found that these lines had been added to his patch without his knowledge!

File: python/tests/test_highlevel.py (hunk 7 above)
Comment (szhan) [reply python/tests/test_highlevel.py:4303]: NVM. I think I misunderstood ``test_bad_position()``.

File: python/tests/test_highlevel.py [hunk 10]
```diff
@@ -4261,3 +4264,79 @@ def test_macs_error(self):
...
+
+    @pytest.mark.parametrize("position", [0.5, 6.2, 8.3])
+    def test_position_continuous_coordinates(self, position):
+        ts = self.get_example_ts_continuous_coordinates()
+        site = ts.site(position=position)
+        assert site.position == position
+
+    @pytest.mark.parametrize("position", [0, 2.999999999, 5.000000001, 9])
+    def test_position_not_found(self, position):
+        with pytest.raises(ValueError):
```
Comment (benjeffery) [inline python/tests/test_highlevel.py:4307]: For all these it is good to add a `match=` to check the error is the right one.

**Review by benjeffery — COMMENTED** [review]
Getting there! One good idea to add `match=` to all the exception testers.

**Comment by benjeffery** [comment]
Not sure what is going on with coverage here as those states must be getting hit. Locally I hit them if I set a breakpoint, but my coverage shows them as uncovered too.

**Comment by szhan** [comment]
Jerome figured it out! I just had to rebase. It seems that codecov didn't like that my branch was based on an older version of tskit.

File: python/tskit/trees.py [hunk 11]
```diff
@@ -5209,13 +5209,33 @@ def mutation(self, id_):
...
+            if len(position.shape) > 0:
+                raise ValueError("Position must be provided as a scalar value.")
+            if position < 0 or position >= self.sequence_length:
+                raise ValueError(
+                    "Position is beyond the coordinates defined by sequence length."
+                )
+            site_pos = self.tables.sites.position  # Sorted in ascending order
+            id_ = site_pos.searchsorted(position)
+            if id_ >= len(site_pos) or site_pos[id_] != position:
+                raise ValueError(f"There is no site at position {position}.")
```
Comment (jeromekelleher) [inline python/tskit/trees.py:5238]: I wonder if this should be a KeyError @benjeffery?

**Review by jeromekelleher — APPROVED** [review]
LGTM!

File: python/tskit/trees.py (hunk 11 above)
Comment (benjeffery) [reply python/tskit/trees.py:5238]: Yes, probably should be, after thinking about it.

Comment (jeromekelleher) [reply python/tskit/trees.py:5238]: Actually no, I take it back. list().index raises a ValueError when the value isn't found, and I guess this is equivalent.

**Pull request merged**